text/0041-shutdown.md

```diff
@@ -0,0 +1,135 @@
++ Feature name: `shutdown`
++ Start date: 2020-07-03
++ RFC PR: [iotaledger/bee-rfcs#41](https://github.com/iotaledger/bee-rfcs/pull/41)
++ Bee issue: [iotaledger/bee#96](https://github.com/iotaledger/bee/issues/96)
+
+# Summary
+
+This RFC proposes a unified way of achieving a graceful shutdown of arbitrarily many async workers and their utilized resources in the Bee framework.
+
+# Motivation
+
+Rust provides one of the most powerful, performant, and still safest abstractions to deal with *concurrency*. You can read about Rust's concurrency story [here](https://doc.rust-lang.org/book/ch16-00-concurrency.html).
+
+One problem that arises when dealing with many asynchronous tasks within a system is to ensure that all of them terminate gracefully once an agreed upon shutdown signal was issued. An example for this is when a software intercepts the press of `^C` in the terminal, and then executes specialized instructions to prevent data corruption before finally exiting the application.
+
+This feature solves this problem by introducing a type called `Shutdown`, that allows to register ansynchronous workers, and `execute` a shutdown in a 3-step process:
+
+1. Send a shutdown signal to all registered workers;
+2. Wait for all the workers to terminate;
+3. Perform final actions (like flushing buffers to databases, saving files, etc.)
+
+# Detailed design
+
+As the shutdown signal itself contains no data, the zero-sized unit type `()` is used to model it. Furthermore, as the synchronizing mechanism, this RFC proposes the use of `oneshot` channels from the `futures` crate. Those channels are consumed, i.e. destroyed, when calling `send` once; this way Rust ensures at compile-time, that no further redundant shutdown signals can be sent to the same listener.
+
+To make the code more readable and less verbose, this RFC proposes the following type aliases:
+
+```rust
+// The "notifier" is the origin of the shutdown signal. For each spawned worker a
+// dedicated notifier will be created and kept in a list.
+type ShutdownNotifier = oneshot::Sender<()>;
+
+// The "listener" is the receiver of the shutdown signal. It awaits the shutdown
+// signal as part of the event loop of the corresponding worker.
+type ShutdownListener = oneshot::Receiver<()>;
+
+// For each async worker the async runtime creates a `Future`, that completes when the
+// worker terminates, i.e. shuts down. To make sure, the shutdown is complete, one has
+// to store those `Future`s in a list, and await all of them iteratively before exiting
+// the program.
+type WorkerShutdown = Box<dyn Future<Output = Result<(), WorkerError>> + Unpin>;
+
+// Before shutting down a system, workers oftentimes need to perform one or more final
+// actions to prevent data loss and/or data corruption. Therefore each worker can register
+// an arbitrary amount of actions, that will be executed as the final step of the shutdown
+// process.
+type Action = Box<dyn Fn()>;
```

Comment:
Can this be extended to `FnOnce()`?

Reply:
I think so. :+1: Each action is supposed to be called only once.